### ISSUE TITLE
A new thread is created on every request when @EnableRedisHttpSession is used.

### ISSUE THREAD

**Issue by tsachev** [spring-projects/spring-session#314]
`RedisHttpSessionConfiguration` creates `RedisMessageListenerContainer` which by
default uses `SimpleAsyncTaskExecutor` to dispatch messages, which means a
new thread is created for every single message received by Redis.
`RedisOperationsSessionRepository` (previously `SessionMessageListener`) is
subscribed to `__keyevent@*:del` pattern.
`RedisSessionExpirationPolicy` sends a `del` event for expirations key,
every time the session is saved (those can be reduced sometimes, but I will discuss this in another issue).
And though those events are ignored by the listener, because it does not care about
expirations keys (`spring:session:expirations:XXXX`) a thread is still created.

So the effect is that a new thread is created for every request.
If the subscription pattern cannot be more specific (i.e. somehow specify that we do not want events for expirations),
then more sensible executors should be configured on the listener container.

`RedisMessageListenerContainer` has `subscriptionExecutor` and `taskExecutor`.
`subscriptionExecutor` defaults to `taskExecutor` if unset, and `taskExecutor` defaults to the bad `SimpleAsyncTaskExecutor`.
It is worth nothing that `subscriptionExecutor` cannot be bounded
thread pool with size < 2, because it starts a long running thread.

Here is one possible work-araound

```
@Configuration
public class Config extends RedisHttpSessionConfiguration {
  @Bean
  public TaskExecutor sessionRedisListenerContainerTaskExecutor() {
    ThreadPoolTaskExecutor taskExecutor = new ThreadPoolTaskExecutor();
    taskExecutor.setThreadNamePrefix("SessionRedisListenerContainer-");
    taskExecutor.setMaxPoolSize(4);
    taskExecutor.setKeepAliveSeconds(serverProperties.getSession().getTimeout());
    return taskExecutor;
  }

  @Bean
  @Override
  public RedisMessageListenerContainer redisMessageListenerContainer(
                  RedisConnectionFactory connectionFactory, RedisOperationsSessionRepository messageListener) {
    RedisMessageListenerContainer container = super.redisMessageListenerContainer(connectionFactory, messageListener);
    container.setTaskExecutor(sessionRedisListenerContainerTaskExecutor());
    return container;
  }
}
```

But probably creating a `SimpleAsyncTaskExecutor` for `subscriptionExecutor` and
bounded thead pool for `taskExecutor` with configurable `maxThreads` is better.

**Comment by rwinch** [comment]
@tsachev Thanks for the detailed report!

Since you appear to have an idea on how to resolve this, would you like to submit a PR? If not, I will be sure to work on this for 1.1 RC1

**Comment by ptahchiev** [comment]
+1 on this

**Comment by rwinch** [comment]
@tsachev Just taking a more detailed look at this.

I'm not sure we can know what the best default is.If we do know a better default, it should probably go in `RedisMessageListenerContainer`.

Instead, I propose that we provide an easy way to override the executors by defining beans of a specific name (i.e. `springSessionRedisTaskExecutor` and `springSessionRedisSubscriptionExecutor `).

What are your thoughts?

**Comment by christophstrobl** [comment]
@rwinch @tsachev `RedisMessageListenerContainer` is unlikely to change its defaults. 
+1 for the override option.

**Comment by tsachev** [comment]
In this case I think that overriding the task executor should be as easy as possible, i.e. +1.

**Comment by rwinch** [comment]
@tsachev Thanks for the response! Any chance you could submit a PR with the ability to provide these beans by default? A few things we would want to ensure

* [ ] The beans are optional
* [ ] We have tests
* [ ] We add documentation on overriding the values just below http://docs.spring.io/spring-session/docs/1.1.0.RC1/reference/html5/#custom-redisserializer

**Comment by tsachev** [comment]
@rwinch OK I will try to come up with PR next week.

**Comment by rwinch** [comment]
Thanks for the fast turnaround @tsachev! This is now merged into master and will be part of 1.1.0
